sdk/projects/azure-projects/azure/projects/_resource.py

```diff
@@ -0,0 +1,632 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+from __future__ import annotations
+
+import inspect
+from itertools import product
+from collections import defaultdict
+import os
+from typing import (
+    Literal,
+    Mapping,
+    MutableMapping,
+    Tuple,
+    Optional,
+    Type,
+    Union,
+    Dict,
+    List,
+    Any,
+    Sequence,
+    Protocol,
+    AsyncContextManager,
+    ContextManager,
+    overload,
+    runtime_checkable,
+    TYPE_CHECKING,
+)
+from typing_extensions import Self, NamedTuple, TypedDict, TypeVar
+
+from dotenv import dotenv_values
+
+from azure.core.settings import _unset
+
+from .resources._identifiers import ResourceIdentifiers
+from ._utils import (
+    run_coroutine_sync,
+    resolve_properties,
+    find_last_resource_match,
+    find_resource_group,
+)
+from ._setting import StoredPrioritizedSetting
+from ._bicep.expressions import (
+    Output,
+    Parameter,
+    ResourceSymbol,
+    ResourceGroup as RGSymbol,
+)
+from ._bicep.utils import clean_name
+
+if TYPE_CHECKING:
+    from azure.core.credentials import SupportsTokenInfo
+    from azure.core.credentials_async import AsyncSupportsTokenInfo
+
+    from .resources.resourcegroup import ResourceGroup
+    from .resources._extension import RoleAssignment
+    from ._component import AzureInfrastructure
+
+
+def _load_dev_environment(name: Optional[str] = None, label: Optional[str] = None) -> Dict[str, Optional[str]]:
+    azd_dir = os.path.join(os.getcwd(), ".azure")
+    if name and not name.endswith(".env"):
+        if not os.path.isdir(azd_dir):
+            return {}
+        env_name = name + (f"-{label}" if label else "")
+        env_path = os.path.join(azd_dir, env_name, ".env")
+    elif name:
+        env_path = name
+    else:
+        script_name = os.path.splitext(os.path.basename(inspect.stack()[-1][0].f_code.co_filename))[0]
+        script_env = os.path.join(azd_dir, script_name, ".env")
+        if os.path.isdir(azd_dir) and os.path.isfile(script_env):
+            env_path = script_env
+        else:
+            env_path = ".env"
+    values = dotenv_values(env_path)
+    return values
+
+
+def _build_envs(services: List[str], attributes: List[str]) -> List[str]:
+    all_vars = product(services, attributes)
+    return ["_".join(["AZURE"] + list(var)).upper() for var in all_vars]
+
+
+@runtime_checkable
+class SyncClient(Protocol, ContextManager["SyncClient"]):
+    def close(self) -> None: ...
+
+
+@runtime_checkable
+class AsyncClient(Protocol, AsyncContextManager["AsyncClient"]):  # pylint: disable=async-client-bad-name
+    async def close(self) -> None: ...
+
+
+class FieldType(NamedTuple):
```

Comment:
We got hurt by NamedTuple in the past because that doesn't allow new property. That's why we have a new token protocol for instance. Is this one safe?

Reply:
Good call - though none of the NamedTuples are exposed int he public API - they purely to streamline and type hint the internal implementation. So modifying them in breaking ways should not be a problem.